src/ts/components/settingspanel.ts

```diff
@@ -72,28 +103,132 @@ export class SettingsPanel extends Container<SettingsPanelConfig> {
       this.onHide.subscribe(() => {
         // Clear timeout when hidden from outside
         this.hideTimeout.clear();
+        // Reset navigation
+        this.resetNavigation();
       });
     }
 
-    // Fire event when the state of a settings-item has changed
-    let settingsStateChangedHandler = () => {
+    // pass event from root page through for backwards compatibility
+    this.getRootPage().onSettingsStateChanged.subscribe(() => {
       this.onSettingsStateChangedEvent();
+    });
 
-      // Attach marker class to last visible item
-      let lastShownItem = null;
-      for (let component of this.getItems()) {
-        component.getDomElement().removeClass(this.prefixCss(SettingsPanel.CLASS_LAST));
-        if (component.isShown()) {
-          lastShownItem = component;
-        }
-      }
-      if (lastShownItem) {
-        lastShownItem.getDomElement().addClass(this.prefixCss(SettingsPanel.CLASS_LAST));
+    this.updateActivePageClass();
+  }
+
+  private updateActivePageClass(): void {
+    this.getPages().forEach((page: SettingsPanelPage, index) => {
+      if (index === this.activePageIndex) {
+        page.getDomElement().addClass(SettingsPanel.ACTIVE_PAGE_CLASS);
+      } else {
+        page.getDomElement().removeClass(SettingsPanel.ACTIVE_PAGE_CLASS);
       }
+    });
+  }
+
+  setActivePageIndex(index: number): void {
+    const targetPage = this.getPages()[index];
+    if (targetPage !== undefined) {
+      this.animateNavigation(targetPage);
+      this.activePageIndex = index;
+      this.navigationStack.push(targetPage);
+      this.updateActivePageClass();
+      targetPage.onActiveEvent();
+    }
+  }
+
+  setActivePage(page: SettingsPanelPage): void {
+    const index = this.getPages().indexOf(page);
+    this.setActivePageIndex(index);
+  }
+
+  popToRootSettingsPanelPage(): void {
+    this.resetNavigation();
+  }
+
+  popSettingsPanelPage() {
+    // pop one navigation item from stack
+    const currentPage = this.navigationStack.pop(); // remove current page
+    const targetPage = this.navigationStack.slice(-1)[0]; // pick target page without removing it
```

Comment:
i would prefer `this.navigationStack[this.navigationStack.length - 1]` as it does the same but expresses the intent better IMO